src/sinks/datadog/metrics/service.rs

```diff
@@ -189,21 +189,45 @@ impl Service<DatadogMetricsRequest> for DatadogMetricsService {
 
             let request = request
                 .into_http_request(api_key)
-                .context(BuildRequestSnafu)?;
-            let response = client.send(request).await?;
-            let (parts, body) = response.into_parts();
-            let mut body = hyper::body::aggregate(body)
-                .await
-                .context(CallRequestSnafu)?;
-            let body = body.copy_to_bytes(body.remaining());
-            Ok(DatadogMetricsResponse {
-                status_code: parts.status,
-                body,
-                batch_size,
-                byte_size,
-                raw_byte_size,
-                protocol,
-            })
+                .context(BuildRequestSnafu)
+                .map_err(|error| DatadogApiError::HttpError { error })?;
+
+            match client.send(request).await {
+                Ok(response) => match response.status() {
+                    // From https://docs.datadoghq.com/api/latest/logs/:
+                    //
+                    // The status codes answered by the HTTP API are:
+                    // 200: OK (v1)
+                    // 202: Accepted (v2)
+                    // 400: Bad request (likely an issue in the payload
+                    //      formatting)
+                    // 403: Permission issue (likely using an invalid API Key)
+                    // 413: Payload too large (batch is above 5MB uncompressed)
+                    // 5xx: Internal error, request should be retried after some
+                    //      time
+                    StatusCode::BAD_REQUEST => Err(DatadogApiError::BadRequest),
+                    StatusCode::FORBIDDEN => Err(DatadogApiError::Forbidden),
+                    StatusCode::OK | StatusCode::ACCEPTED => {
+                        let (parts, body) = response.into_parts();
+                        let mut body = hyper::body::aggregate(body)
+                            .await
+                            .context(CallRequestSnafu)
+                            .map_err(|error| DatadogApiError::HttpError { error })?;
+                        let body = body.copy_to_bytes(body.remaining());
+                        Ok(DatadogMetricsResponse {
+                            status_code: parts.status,
+                            body,
+                            batch_size,
+                            byte_size,
+                            raw_byte_size,
+                            protocol,
+                        })
+                    }
+                    StatusCode::PAYLOAD_TOO_LARGE => Err(DatadogApiError::PayloadTooLarge),
+                    _ => Err(DatadogApiError::ServerError),
```

Comment:
This should probably boil down to a method on `DatadogApiError` called `from_status` or something, that returns an `Option<DatadogApiError>` so that you can do something like:

```rust
match DatadogApiError::from_status(response.status()) {
    Some(err) => Err(err),
    None => {
        let (parts, body) = response.into_parts();
        ...
    },
}
```

Otherwise we're using a common error type that could theoretically map different status codes to different error variants, which doesn't feel like it fits the goal of a common error type.

Reply:
Alternatively, you could write a method called `from_result` that takes an `http::Response` and has a return value of `Result<http::Response, DatadogApiError>` and then you just roll up all the above into that method... basically only returning `Ok(response)` if you got OK/Accepted, and `Err(...)` otherwise.

Reply:
Makes sense, went with the second approach.